plugins/cpp_metrics/parser/src/cppmetricsparser.cpp

```diff
@@ -101,170 +100,179 @@ bool CppMetricsParser::cleanupDatabase()
+    [&, this](const MetricsTasks<model::CppFunctionMcCabe>& tasks)
   {
-    for (const model::CppFunctionMcCabe& function
-      : _ctx.db->query<model::CppFunctionMcCabe>())
+    util::OdbTransaction {_ctx.db} ([&, this]
     {
-      // Skip functions that were included from external libraries.
-      if (!cc::util::isRootedUnderAnyOf(_inputPaths, function.filePath))
-        continue;
-
-      model::CppAstNodeMetrics funcMcCabe;
-      funcMcCabe.astNodeId = function.astNodeId;
-      funcMcCabe.type = model::CppAstNodeMetrics::Type::MCCABE;
-      funcMcCabe.value = function.mccabe;
-      _ctx.db->persist(funcMcCabe);
-    }
+      for (const model::CppFunctionMcCabe& param : tasks)
+      {
+        model::CppAstNodeMetrics funcMcCabe;
+        funcMcCabe.astNodeId = param.astNodeId;
+        funcMcCabe.type = model::CppAstNodeMetrics::Type::MCCABE;
+        funcMcCabe.value = param.mccabe;
+        _ctx.db->persist(funcMcCabe);
+      }
+    });
   });
 }
 
 void CppMetricsParser::functionBumpyRoad()
 {
-  util::OdbTransaction {_ctx.db} ([&, this]
+  // Calculate the bumpy road metric for all types on parallel threads.
+  parallelCalcMetric<model::CppFunctionBumpyRoad>(
+    "Bumpy road complexity",
+    _threadCount * 5,// number of jobs; adjust for granularity
+    getFilterPathsQuery<model::CppFunctionBumpyRoad>(),
+    [&, this](const MetricsTasks<model::CppFunctionBumpyRoad>& tasks)
   {
-    for (const model::CppFunctionBumpyRoad& function
-      : _ctx.db->query<model::CppFunctionBumpyRoad>())
+    util::OdbTransaction {_ctx.db} ([&, this]
     {
-      // Skip functions that were included from external libraries.
-      if (!cc::util::isRootedUnderAnyOf(_inputPaths, function.filePath))
-        continue;
-
-      const double dB = function.bumpiness;
-      const double dC = function.statementCount;
-      const bool empty = function.statementCount == 0;
-
-      model::CppAstNodeMetrics metrics;
-      metrics.astNodeId = function.astNodeId;
-      metrics.type = model::CppAstNodeMetrics::Type::BUMPY_ROAD;
-      metrics.value = empty ? 1.0 : (dB / dC);
-      _ctx.db->persist(metrics);
-    }
+      for (const model::CppFunctionBumpyRoad& function : tasks)
+      {
+        const double dB = function.bumpiness;
+        const double dC = function.statementCount;
+        const bool empty = function.statementCount == 0;
+
+        model::CppAstNodeMetrics metrics;
+        metrics.astNodeId = function.astNodeId;
+        metrics.type = model::CppAstNodeMetrics::Type::BUMPY_ROAD;
+        metrics.value = empty ? 1.0 : (dB / dC);
+        _ctx.db->persist(metrics);
+      }
+    });
   });
 }
 
 void CppMetricsParser::lackOfCohesion()
 {
-  util::OdbTransaction {_ctx.db} ([&, this]
+  // Calculate the cohesion metric for all types on parallel threads.
+  parallelCalcMetric<model::CohesionCppRecordView>(
+    "Lack of cohesion",
+    _threadCount * 25,// number of jobs; adjust for granularity
```

Comment:
How were these magic numbers decided?

Reply:
Extracted them at least to const members. See fc86a64c0fa4e4299814b9d48a98378c00522309.